### PR TITLE
[Merged by Bors] - Added boa examples

### LINKED ISSUE #446: Add examples
**Current situation**
We currently have no code examples to use the API externally both in the API documentation and in the nonexistent `examples` folder.

**What do we need?**
As mentioned in #445, we should be creating a new API, but in any case, we would need to create examples for some common use cases of Boa. I'm extracting here part of that discussion:

> When a user wants to use Boa for their projects, I can imagine the following use cases:
>  - They want to interpret JavaScript from a file or a stored String and get the result.
>  - They want to interpret JavaScript that is coming from a socket with a `Read` interface and get the result by streaming the bytes.
>  - They want to extend our engine by providing new global functions and objects, so that with the JavaScript they can either perform more rich things in their setup (for example by providing an API for a browser window), or because they want to add JavaScript helpers for common actions in their setup. This also includes things as being the scripting language of a complex platform.
>  - They want to change the `global` object, for example, because they want it to be the `window` or a `frame`.

I think we need at least one example of each, that gets tested on each commit, in the `examples` directory of the `boa` crate. Then, I would add parts of those examples to the library-level documentation, in order to make life easier for our users.

### PR DESCRIPTION
**Pull request by jasonwilliams**
Added boa examples as per issue #446
Overtaken https://github.com/boa-dev/boa/pull/634

Somehow screwed that branch up by rebasing it and losing access
pings @elasmojs 

This Pull Request fixes/closes #446 .

### CODE REVIEW

**Review by RageKnify — CHANGES_REQUESTED** [review]
Some minor things in the `Cargo.toml` files.

File: examples/src/loadfile.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+pub fn run(){
+    let js_file_path = "./scripts/helloworld.js";
+    let buffer = read_to_string(js_file_path);
+
+    if buffer.is_err(){
+        println!("Error: {}", buffer.unwrap_err());
+        return;
+    }
```
Comment (RageKnify) [inline examples/src/loadfile.rs:16]: Make it more idiomatic.
```suggestion
    if let Err(msg) = buffer {
        println!("Error: {}", msg);
        return;
    }
```

File: examples/src/modulehandler.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    //Adding custom implementation that mimics 'require'
+    let requirefn = Function::builtin(Vec::new(), require);
+    ctx.global_obj.set_field("require", Value::from_func(requirefn));
+
+    //Addming custom object that mimics 'module.exports'
+    let moduleobj = Value::new_object(Some(&ctx.global_obj));
+    moduleobj.set_field("exports", Value::from(" "));
+    ctx.global_obj.set_field("module", moduleobj);
+
+    //Instantiating the engien with the execution context
```
Comment (RageKnify) [inline examples/src/modulehandler.rs:33]:
```suggestion
    //Instantiating the engine with the execution context
```

**Review by RageKnify — CHANGES_REQUESTED** [review]
1 typo and a change to make the Rust code more idiomatic.

File: Cargo.toml [hunk 3]
```diff
@@ -5,6 +5,7 @@ members = [
     "boa_wasm",
     "boa_tester",
     "boa_unicode",
+    "boa_examples",
```
Comment (RageKnify) [inline Cargo.toml:8]:
```suggestion
    "examples",
```
I think this line needs to have the name of the directory at the top of the repository. The other option is changing the name of the directory from `examples` to `boa_examples`.

**Comment by Razican** [comment]
Hello @jasonwilliams how is this going?

**Comment by jasonwilliams** [comment]
OK, from what i remember it was just the change @RageKnify pointed out.
Will try to get back on this this week

**Comment by Razican** [comment]
I have updated the most simple examples, but the rest require some extra work.

File: boa_examples/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    println!("\r\n");
+
+    //example that enhances the global object with custom values, objects, functions
+    enhancedglobal::run();
+    println!("\r\n");
+
+    //example that implements a custom module handler which mimics (require / module.exports) pattern
+    modulehandler::run();
+    println!("\r\n");
+}
```
[the remark this answers is not in the document]
Comment (jasonwilliams) [reply boa_examples/src/main.rs]: done

**Review by RageKnify — APPROVED** [review]

**Review by Razican — COMMENTED** [review]
Looks good to me, I think these examples will be useful for our users. The only thing I'm missing is a small description on the top of each example with module level comments (`!//`) explaining what the example is supposed to do.

Also, I would say we don't have an example that changes the global object, right? Do we want to have this? Is it possible with the current API?

**Comment by jasonwilliams** [comment]
> Looks good to me, I think these examples will be useful for our users. The only thing I'm missing is a small description on the top of each example with module level comments (`//!`) explaining what the example is supposed to do.
> 
> Also, I would say we don't have an example that changes the global object, right? Do we want to have this? Is it possible with the current API?

How do you mean change the global object? Do you mean to add properties to it?

**Comment by Razican** [comment]
> How do you mean change the global object? Do you mean to add properties to it?

I meant using "window" as a global object for example, but I guess that's just an alias, right? And then adding some attributes to it.

**Comment by Razican** [comment]
@jasonwilliams do you think we could have this ready for 0.14? It would be nice to have new examples here.

**Comment by Razican** [comment]
I will look at this during the week, I would love to merge it asap :)

**Comment by jasonwilliams** [comment]
`gc` needed to be added as a dependency, not sure why when using boa_gc. Any ideas @Razican ?

**Comment by Razican** [comment]
> `gc` needed to be added as a dependency, not sure why when using boa_gc. Any ideas @Razican ?

This is due to the way deriving works for that crate. Maybe it can be fixed with a PR to the gc crate

**Comment by jasonwilliams** [comment]
@jedel1043 you happy with this?

**Review by jedel1043 — APPROVED** [review]
> @jedel1043 you happy with this?

Yep :)

**Pull request closed without merging**